### ISSUE TITLE
mkExtraTvbs is needlessly complex

### ISSUE THREAD

**Issue by RyanGlScott** [goldfirere/th-desugar#73]
While pondering #68, I was reminded of just how horrifically complicated the `mkExtraTvbs` function in `Language.Haskell.TH.Desugar.Core` is:

https://github.com/goldfirere/th-desugar/blob/4ca98c6492015e6ad063d3ad1a2ad6c4f0a56837/Language/Haskell/TH/Desugar/Core.hs#L1004-L1032

This is really, really gross. Why are things this way? It turns out this mess started festering due to a innocent-enough looking test called `dectest9`:

https://github.com/goldfirere/th-desugar/blob/9f4be6dae9a083cccfb1faeeca541afca7a6844d/Test/Splices.hs#L355-L358

Our test suite tries very hard to ensure that when we desugar `Dec9` directly into a `DDec`, it matches what we would get had we reified `Dec9` and then desugared it. In fact, it likely tries _too_ hard.

There's an annoying quirk of GHC that when we reify a datatype with an explicit kind signature like `DDec`, it translates the kind signature into fresh type variables. For example, `Dec9` would likely get reified as:

```haskell
data family Dec9 a (b :: * -> *) (c :: *) -- c is a gensym'd type variable
```

However, this means that we have to do the exact same thing in `th-desugar`, lest the test suite fail. A consequence of this is that we have to _guess the exact same variable names that GHC gensyms_, which is why `mkExtraTvbs` is so complicated in the first place. Ugh.

To make things even worse, @goldfirere discovered that `dectest9` is not robust in the face of recompilation, as rebuilding can sometimes sway GHC to gensym different variable names. Thus, not only is `mkExtraTvbs` overengineered, it doesn't even accomplish the task it was built for!

All of this seems to suggest that our testing should be robust against this sort of behavior. I propose the following change: when testing equality of `DDec`s, we ignore the names of the type variable binders in data types. This isn't as annoying to implement as it sounds, as we're already using a `Data.Data`–based API, so the only change we'd need to the test suite is:

```diff
diff --git a/Test/Splices.hs b/Test/Splices.hs
index 894abb9..c415970 100644
--- a/Test/Splices.hs
+++ b/Test/Splices.hs
@@ -96,10 +96,19 @@ dropTrailing0s = everywhere (mkT (mkName . frob . nameBase))
       | head str == 'R' = str
       | otherwise       = dropWhileEnd isDigit str
 
-eqTH :: (Data a, Show a) => a -> a -> Bool
-eqTH a b = show (unqualify a) == show (unqualify b)
+eqTH :: (Data a, Eq a) => a -> a -> Bool
+eqTH a b = extQ2 (==) eqDDec (unqualify a) (unqualify b)
+  where
+    extQ2 :: (Typeable a, Typeable b)
+          => (a -> a -> q)
+          -> (b -> b -> q)
+          -> a -> a -> q
+    extQ2 f g x y = maybe (f x y) (uncurry g) ((,) <$> cast x <*> cast y)
+
+    eqDDec :: DDec -> DDec -> Bool
+    eqDDec = ...
 
-eqTHSplice :: (Data a, Show a) => a -> a -> Q Exp
+eqTHSplice :: (Data a, Eq a) => a -> a -> Q Exp
 eqTHSplice a b =
   if a `eqTH` b
   then [| True |]
```

(This would require that we derive `Eq` instances for everything in `Language.Haskell.TH.Desugar.Core`, but to be honest, I'm surprised we're not doing this already!)

Once that is done, I think we could replace an enormous chunk of `mkExtraTvbs` with `newName`/`newUniqueName`, which would greatly simplify it.

**Comment by RyanGlScott** [comment]
Actually, I take back what I said about `Eq` instances, since it turns out my plan in https://github.com/goldfirere/th-desugar/issues/73#issue-296123088 wasn't fully baked. :)

There's one big problem I forgot about: data family instances. Here, we don't have type variable binders, but rather type patterns (e.g., `data instance F a b a :: Type -> Type where MkF :: a -> b -> F a b a c`). How should we normalize this? We could adopt the same approach of just ignoring all type variable names when checking for equality, but this is perhaps too loosey-goosey. After all, under this notion of equality, we would treat `data instance F a b a :: Type -> Type` and `data instance F a b c :: Type -> Type` as being the same, when they really aren't.

Another approach would be to extract all of the variable names from the data family instance, map them to a list of variable names that are generated deterministically, and _then_ check for equality. But that would introduce a large amount of complexity since we check for equality in the `dectest`s on `Dec`s, not `DDec`s, so at the very minimum we'd need to come up with a variant of `fvDType` that works over `Type`s instead of `DType`s... blech.

-----

However, there is another way to work around this problem that I should have realized earlier. Recall that this issue happens in the first place because when we roundtrip through `th-desugar`, we first convert `:: * -> *` to `(c :: *) :: *` (where `c` is gensym'd). But if we implement https://github.com/goldfirere/th-desugar/issues/68, then this would never happen in the first place, since we'd properly desugar `:: * -> *` to `:: * -> *`. Thus, roundtripping through `th-desugar` would no longer need to gensym anything at all, which sidesteps this issue completely.

This wouldn't be a perfect solution, of course. Template Haskell didn't have the ability to represent data type return kinds until GHC 8.0, so this wouldn't work on older GHCs. And to make things more interesting, if we implemented https://github.com/goldfirere/th-desugar/issues/68, then the `dectest`s would probably need to be overhauled anyway, since roundtripping through `th-desugar` would convert all data types to GADT syntax, so something like `dectest1`:

https://github.com/goldfirere/th-desugar/blob/4a962db68878f1a03ecfa4d164b48b119e47b5b2/Test/Splices.hs#L315

Would no longer pass, since `th-desugar` would convert that to:

```haskell
data Dec1 :: Type where
  Foo :: Dec1
  Bar :: Int -> Dec1
```

Perhaps all of this is an acceptable price to pay, though. We could only test data types with explicit return kinds on GHC 8.0 or later, and we could change the `dectest`s to all use GADT syntax. It would be something we'd have to be aware of, though.

**Comment by goldfirere** [comment]
These are all reasonable designs, in my view. I'm afraid I don't have a concrete opinion on which is best, but there's nothing here I object to. The design space here is non-trivial.

**Comment by RyanGlScott** [comment]
I've submitted #74 to address this, which redesigns `th-desugar`'s representation of data types such that the gensym algorithm in `mkExtraTvbs` is no longer needed (and can thus be replaced with `newName`). Even better, I didn't need to CPP out any tests on old versions of GHC. Hooray!
